presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayBlockBuilder.java

```diff
@@ -190,6 +190,16 @@ public ArrayBlock build()
     }
 
     @Override
+    public void reset(BlockBuilderStatus blockBuilderStatus)
+    {
+        this.blockBuilderStatus = requireNonNull(blockBuilderStatus, "blockBuilderStatus is null");
+        currentEntrySize = 0;
+        valueIsNull = new DynamicSliceOutput(valueIsNull.size());
```

Comment:
instead of continuing to expand the BlockBuilders with the max size seen so far, we should expand based on the last known used value with some skew (e.g. 1.25)